frontend/src/features/infrastructure/InfrastructurePage.tsx

```diff
@@ -1,3 +1,70 @@
+import { useEffect, useState } from 'react'
+import { useSearchParams } from 'react-router-dom'
+import { useStatus } from '../../api/status/use-status'
+import { Tabs } from '../../components/Tabs'
+import { ControllersList } from './ControllersList'
+import { DeploymentsList } from './DeploymentsList'
+import { RoutesList } from './RoutesList'
+import { RunnersList } from './RunnersList'
+
 export const InfrastructurePage = () => {
-  return <>Infrastructure</>
+  const status = useStatus()
+  const [searchParams, setSearchParams] = useSearchParams()
+
+  const [tabs, setTabs] = useState([
+    { name: 'Controllers', id: 'controllers' },
+    { name: 'Runners', id: 'runners' },
+    { name: 'Deployments', id: 'deployments' },
+    { name: 'Routes', id: 'routes' },
+  ])
+
+  useEffect(() => {
+    if (!status.data) {
+      return
+    }
+
+    setTabs((prevTabs) =>
+      prevTabs.map((tab) => {
+        switch (tab.id) {
+          case 'controllers':
+            return { ...tab, count: status.data.controllers.length }
+          case 'runners':
+            return { ...tab, count: status.data.runners.length }
+          case 'deployments':
+            return { ...tab, count: status.data.deployments.length }
+          case 'routes':
+            return { ...tab, count: status.data.routes.length }
+          default:
+            return tab
+        }
+      }),
+    )
+  }, [status.data])
+
+  const handleTabClick = (tabId: string) => {
+    setSearchParams({ tab: tabId })
+  }
+
+  const currentTab = searchParams.get('tab') || tabs[0].id
```

Comment:
What do you think about making the tab part of the path? Since these tabs are basically sub-pages of the Infrastructure page, that feels a little more canonical. Query params work too, though.

Reply:
Yeah it's a good question! I think for now I might leave the params and as I build this out re-evaluate the best solution here. Good call out!

Reply:
sounds good!